examples_tests/48.ArithmeticUnitTest/main.cpp

```diff
@@ -19,25 +20,71 @@ enum TestOperation {
 	TO_MAX
 };
 
-template<typename T, typename OP, uint32_t subgroupSize>
-T emulatedSubgroupReduction(const T* data, const uint32_t globalInvocationIndex)
+template<typename T> T and(T left, T right) { return left & right; }
+template<typename T> T xor(T left, T right) { return left ^ right; }
+template<typename T> T or(T left, T right) { return left | right; }
+template<typename T> T add(T left, T right) { return left + right; }
+template<typename T> T mul(T left, T right) { return left * right; }
+template<typename T> T min(T left, T right) { return std::min(left, right); }
+template<typename T> T max(T left, T right) { return std::max(left, right); }
+
+typedef uint32_t(*arithmeticFuncPtr)(uint32_t, uint32_t);
+arithmeticFuncPtr arrayofFunctions[] = {
+ and<uint32_t>,
+ xor<uint32_t>,
+ or <uint32_t>,
+ add<uint32_t>,
+ mul<uint32_t>,
+ min<uint32_t>,
+ max<uint32_t> 
+};
+
+//subgroup method emulations on the CPU, to verify the results of the GPU methods
+template<typename T>
+T emulatedSubgroupReduction(const T* data, const uint32_t globalInvocationIndex, uint32_t subgroupSize, arithmeticFuncPtr pFunc)
 {
 	auto subgroupID = globalInvocationIndex / subgroupSize;
-	//auto subgroupOffset = data + subgroupID * subgroupSize;
 	T retval = data[0];
 	for (auto i = 1u; i < subgroupSize; i++)
-		retval = OP(retval, data[i]);
+		retval = pFunc(retval, data[i]);
```

Comment:
its wront you need the subgroup offset to data 🤦

Reply:
so it actually had a use then...

Reply:
everything is fine, but the variable name is misleading